src/frontend/Semantic_error.ml

```diff
@@ -152,6 +156,68 @@ module TypeError = struct
           (String.concat ~sep:"" lines)
           Fmt.(list UnsizedType.pp ~sep:comma)
           arg_tys
+    | IllTypedVariadicODE (name, arg_tys, args) ->
+        let types x = List.map ~f:(fun (_, t) -> t) x in
+        let optional_tol_args =
+          if Stan_math_signatures.is_variadic_ode_tol_fn name then
+            List.map
+              ~f:(fun (_, y) -> y)
+              Stan_math_signatures.variadic_ode_tol_arg_types
+          else []
+        in
+        let args_types = List.map ~f:(fun (_, t) -> t) args in
+        let generate_ode_sig =
+          [ UnsizedType.UFun
+              ( Stan_math_signatures.variadic_ode_mandatory_fun_args @ args
+              , ReturnType Stan_math_signatures.variadic_ode_fun_return_type )
+          ]
+          @ types Stan_math_signatures.variadic_ode_mandatory_arg_types
+          @ optional_tol_args @ args_types
+        in
+        let variadic_ode_generic_signature =
+          let optional_tol_args =
+            if Stan_math_signatures.is_variadic_ode_tol_fn name then
+              types Stan_math_signatures.variadic_ode_tol_arg_types
+            else []
+          in
+          let nth_fun_arg n =
+            List.nth_exn
+              (types Stan_math_signatures.variadic_ode_mandatory_fun_args)
+              n
+          in
+          let nth_arg n =
+            List.nth_exn
+              (types Stan_math_signatures.variadic_ode_mandatory_arg_types)
```

Comment:
this might recompute the result of `(types ...)` every time it's called - very tiny issue, but FWIW one alternative way of writing it might be something like

```
let arg0 :: arg1 :: arg2 :: _ = types ...
```

Reply:
Thanks. But this way I get a warning for a not-exhaustive pattern matching that I am unable to fix: 
File "src/frontend/Semantic_error.ml", line 192, characters 10-471:
Error (warning 8): this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
(_::_::[]|_::[]|[])

Reply:
Ignore the above. Fixed.

Reply:
Oh, yeah, that's unfortunate. You could get around it like:

```
let (arg0, arg1, arg2) = match types ... with
   a0 :: a1 :: a2 :: _ -> (a0, a1, a2)
   _ -> error "oops"
```

but again it's a very small issue that probably doesn't deserve our attention :P

Reply:
Went with:

```ocaml
match
            ( types Stan_math_signatures.variadic_ode_mandatory_arg_types
            , types Stan_math_signatures.variadic_ode_mandatory_fun_args )
          with
          | arg0 :: arg1 :: arg2 :: _, fun_arg0 :: fun_arg1 :: _ ->
              Fmt.strf "(%a, %a, ...) => %a, %a, %a, %a, %a ...\n"
                UnsizedType.pp fun_arg0 UnsizedType.pp fun_arg1 UnsizedType.pp
                Stan_math_signatures.variadic_ode_fun_return_type
                UnsizedType.pp arg0 UnsizedType.pp arg1 UnsizedType.pp arg2
                Fmt.(list UnsizedType.pp ~sep:comma)
                optional_tol_args
          | _ ->
              raise_s
                [%message
                  "This should not happen. Variadic ODE functions have exactly three \
                   mandatory arguments and the function supplied to the variadic ODE \
                   function has exactly two mandatory arguments."]
```

Reply:
Looks good!